### PR TITLE
Fix parsing issue when a backslash as the last character on sudoers file line

### PR DESCRIPTION
**Pull request by iko1**
This PR fixes the bug reported at #7421. I handle the case that a line contains a backslash as last character.
The code aggregates multiple lines within  a "rule_details" column. The aggregation logic starts when a line ends with backslash and end when EOF or the next line doesn't end with a backlash.
The code deletes the backslash character in favor of clearness. 
Any feedback is more than welcome, Thanks!

### CODE REVIEW

File: osquery/tables/system/posix/sudoers.cpp [hunk 1]
```diff
@@ -61,6 +62,20 @@ void genSudoersFile(const std::string& filename,
       continue;
     }
 
+    // if last line contains a backslash as the last character,
+    // treat current line as part of previous line and
+    // append it to appropriate column.
+    if (isLongLine) {
```
Comment (mike-myers-tob) [inline osquery/tables/system/posix/sudoers.cpp:68]: Nit: if following the convention of the rest of this file, variable names are in `snake_case` rather than `camelCase`

File: osquery/tables/system/posix/sudoers.cpp [hunk 2]
```diff
@@ -87,6 +102,11 @@ void genSudoersFile(const std::string& filename,
       continue;
     }
 
+    // Check if a blackslash is the last character on this line.
+    if (!is_include && !is_includedir && line.at(line.size() - 1) == '\\') {
```
Comment (mike-myers-tob) [inline osquery/tables/system/posix/sudoers.cpp:106]: Here and on line 69 above, `line.back()` would be a shorter more readable way to express `line.at(line.size() - 1)`

Comment (iko1) [reply osquery/tables/system/posix/sudoers.cpp]: Fixed

File: osquery/tables/system/posix/sudoers.cpp (hunk 1 above)
Comment (iko1) [reply osquery/tables/system/posix/sudoers.cpp]: changed variable name to snake_case instead of caseCase.

File: osquery/tables/system/posix/sudoers.cpp [hunk 3]
```diff
@@ -61,6 +62,20 @@ void genSudoersFile(const std::string& filename,
...
+    // treat current line as part of previous line and
+    // append it to appropriate column.
+    if (isLongLine) {
+      isLongLine = (line.at(line.size() - 1) == '\\');
+      if (results.back()["rule_details"].empty()) {
+        results.back()["header"].pop_back();
+      } else {
+        results.back()["rule_details"].pop_back();
+      }
+      results.back()["rule_details"].append(line);
```
Comment (Smjert) [inline osquery/tables/system/posix/sudoers.cpp:75]: nit: Prefer storing the result of `results.back()` in a variable first, for instance `const auto& last_line = results.back()` instead of repeating `results.back()`.

File: osquery/tables/system/posix/sudoers.cpp [hunk 4]
```diff
@@ -87,6 +103,11 @@ void genSudoersFile(const std::string& filename,
       continue;
     }
 
+    // Check if a blackslash is the last character on this line.
```
Comment (mike-myers-tob) [inline osquery/tables/system/posix/sudoers.cpp:106]: This detection of a long line occurs after detection of an `include` or a comment, both of which will skip the detection of a long line. So I guess I'm wondering if a file could have lines like:

```
#
# This is a comment that is on a long line ending with a \
this is more comment but a continuation of the above line
# This is another comment line
```

or 

```
#includedir /etc/super/long/path/that/is/broken/by/a\ 
more/long/path/which/ends/on/this/line/sudoers.d
```

Comment (iko1) [reply osquery/tables/system/posix/sudoers.cpp:106]: @mike-myers-tob, No, this syntax is invalid. According of the manual of sudoers: The pound sign (‘#’) is used to indicate a comment or include file or directory directives. the pound sign is up to end-of-line and can't be continued in the next line with backslash. 
Moreover, you can copy both examples into a file and then execute the command: `visudo -c <file>` this is 'check-only' mode,  a supplied sudoers file will be checked for syntax errors. 
spoiler: you'll get a syntax error for your both examples.

Comment (mike-myers-tob) [reply osquery/tables/system/posix/sudoers.cpp:106]: Excellent okay thanks for explaining

**Comment by mike-myers-tob** [comment]
Looks good to me (I reviewed, but don't have the ability approve a PR)

**Review by mike-myers-tob — DISMISSED** [review]

**Review by sharvilshah — COMMENTED** [review]
Hey @iko1, thanks so much for this! Do you think we need to add/update the [`sudoers_test`](https://github.com/osquery/osquery/blob/master/osquery/tables/system/tests/posix/sudoers_tests.cpp) to handle this case?

**Comment by iko1** [comment]
> Hey @iko1, thanks so much for this! Do you think we need to add/update the [`sudoers_test`](https://github.com/osquery/osquery/blob/master/osquery/tables/system/tests/posix/sudoers_tests.cpp) to handle this case?

@sharvilshah, Indeed, the change should be covered in unit test. I'm thinking about adding the following tests:
1. Trivial case when a long line exists in a line that doesn't start with '#'
2. Line that starts with '#' and end with blackslash should be parsed as a single line.

If you have any other edge cases that should be covered, please elaborate them here

**Comment by sharvilshah** [comment]
> > Hey @iko1, thanks so much for this! Do you think we need to add/update the [`sudoers_test`](https://github.com/osquery/osquery/blob/master/osquery/tables/system/tests/posix/sudoers_tests.cpp) to handle this case?
> 
> @sharvilshah, Indeed, the change should be covered in unit test. I'm thinking about adding the following tests:
> 
> 1. Trivial case when a long line exists in a line that doesn't start with '#'
> 2. Line that starts with '#' and end with blackslash should be parsed as a single line.
> 
> If you have any other edge cases that should be covered, please elaborate them here

Those sound great, thanks!

**Review by sharvilshah — APPROVED** [review]
Looks good, thanks
